File: Java/source/src/main/java/com/yitter/core/SnowWorkerM1.java
/*
 * 版权属于：yitter(yitter@126.com)
 * 开源地址：https://gitee.com/yitter/idgenerator
 */
package com.yitter.core;

import com.yitter.contract.ISnowWorker;
import com.yitter.contract.IdGeneratorException;
import com.yitter.contract.IdGeneratorOptions;
import com.yitter.contract.OverCostActionArg;

public class SnowWorkerM1 implements ISnowWorker {

    /**
     * 基础时间
     */
    protected final long BaseTime;

    /**
     * 机器码
     */
    protected final short WorkerId;

    /**
     * 机器码位长
     * （机器码+序列数<=22位）
     */
    protected final byte WorkerIdBitLength;

    /**
     * 自增序列数位长
     * （机器码+序列数<=22位）
     */
    protected final byte SeqBitLength;

    /**
     * 最大序列数（含此值）
     * 超过最大值，就会从MinSeqNumber开始
     */
    protected final int MaxSeqNumber;

    /**
     * 最小序列数（含此值）
     */
    protected final short MinSeqNumber;

    /**
     * 最大漂移次数
     */
    protected final int TopOverCostCount;

    protected final byte _TimestampShift;
    protected final static byte[] _SyncLock = new byte[0];

    protected short _CurrentSeqNumber;
    protected long _LastTimeTick = -1L;
    protected long _TurnBackTimeTick = -1L;
    protected boolean _IsOverCost = false;
    protected int _OverCostCountInOneTerm = 0;
    protected int _GenCountInOneTerm = 0;
    protected int _TermIndex = 0;

    public SnowWorkerM1(IdGeneratorOptions options) {
        WorkerId = options.WorkerId;
        WorkerIdBitLength = options.WorkerIdBitLength == 0 ? 6 : options.WorkerIdBitLength;
        SeqBitLength = options.SeqBitLength == 0 ? 6 : options.SeqBitLength;
        MaxSeqNumber = options.MaxSeqNumber > 0 ? options.MaxSeqNumber : (int) Math.pow(2, SeqBitLength);
        MinSeqNumber = options.MinSeqNumber;
        TopOverCostCount = options.TopOverCostCount;
        BaseTime = options.BaseTime != 0 ? options.BaseTime : 1582136402000L;
        _TimestampShift = (byte) (WorkerIdBitLength + SeqBitLength);
        _CurrentSeqNumber = options.MinSeqNumber;
    }

    private void DoGenIdAction(OverCostActionArg arg) {

    }

    private void BeginOverCostCallBack(long useTimeTick) {
//        if (GenAction == null) {
//            return;
//        }
//
//        DoGenIdAction(new OverCostActionArg(
//                WorkerId,
//                useTimeTick,
//                1,
//                _OverCostCountInOneTerm,
//                _GenCountInOneTerm,
//                _TermIndex));
    }

    private void EndOverCostCallBack(long useTimeTick) {
        if (_TermIndex > 10000) {
            _TermIndex = 0;
        }
//
//        if (GenAction == null) {
//            return;
//        }
//
//        DoGenIdAction(new OverCostActionArg(
//                WorkerId,
//                useTimeTick,
//                2,
//                _OverCostCountInOneTerm,
//                _GenCountInOneTerm,
//                _TermIndex));
    }

    private void TurnBackCallBack(long useTimeTick) {
//        if (GenAction == null) {
//            return;
//        }
//
//        DoGenIdAction(new OverCostActionArg(
//                WorkerId,
//                useTimeTick,
//                8,
//                _OverCostCountInOneTerm,
//                _GenCountInOneTerm,
//                _TermIndex));
    }

    private long NextOverCostId() {
        long currentTimeTick = GetCurrentTimeTick();

        if (currentTimeTick > _LastTimeTick) {
            EndOverCostCallBack(currentTimeTick);

            _LastTimeTick = currentTimeTick;
            _CurrentSeqNumber = MinSeqNumber;
            _IsOverCost = false;
            _OverCostCountInOneTerm = 0;
            _GenCountInOneTerm = 0;

            return CalcId(_LastTimeTick);
        }

        if (_OverCostCountInOneTerm >= TopOverCostCount) {
            EndOverCostCallBack(currentTimeTick);

            _LastTimeTick = GetNextTimeTick();
            _CurrentSeqNumber = MinSeqNumber;
            _IsOverCost = false;
            _OverCostCountInOneTerm = 0;
            _GenCountInOneTerm = 0;

            return CalcId(_LastTimeTick);
        }

        if (_CurrentSeqNumber > MaxSeqNumber) {
            _LastTimeTick++;
            _CurrentSeqNumber = MinSeqNumber;
            _IsOverCost = true;
            _OverCostCountInOneTerm++;
            _GenCountInOneTerm++;

            return CalcId(_LastTimeTick);
        }

        _GenCountInOneTerm++;
        return CalcId(_LastTimeTick);
    }

    private long NextNormalId() throws IdGeneratorException {
        long currentTimeTick = GetCurrentTimeTick();

        if (currentTimeTick > _LastTimeTick) {
            _LastTimeTick = currentTimeTick;
            _CurrentSeqNumber = MinSeqNumber;

            return CalcId(_LastTimeTick);
        }

        if (_CurrentSeqNumber > MaxSeqNumber) {
            BeginOverCostCallBack(currentTimeTick);

            _TermIndex++;
            _LastTimeTick++;
            _CurrentSeqNumber = MinSeqNumber;
            _IsOverCost = true;
            _OverCostCountInOneTerm++;
            _GenCountInOneTerm = 1;

            return CalcId(_LastTimeTick);
        }

        if (currentTimeTick < _LastTimeTick) {
            if (_TurnBackTimeTick < 1) {
                _TurnBackTimeTick = _LastTimeTick - 1;
            }

            try {
                Thread.sleep(10);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }

            TurnBackCallBack(_TurnBackTimeTick);

            return CalcTurnBackId(_TurnBackTimeTick);
        }

        return CalcId(_LastTimeTick);
    }

    private long CalcId(long useTimeTick) {
        long result = ((useTimeTick << _TimestampShift) +
                ((long) WorkerId << SeqBitLength) +
                (int) _CurrentSeqNumber);

        _CurrentSeqNumber++;
        return result;
    }

    private long CalcTurnBackId(long useTimeTick) {
        long result = ((useTimeTick << _TimestampShift) +
                ((long) WorkerId << SeqBitLength) + 0);

        _TurnBackTimeTick--;
        return result;
    }

    protected long GetCurrentTimeTick() {
        long millis = System.currentTimeMillis();
        return millis - BaseTime;
    }

    protected long GetNextTimeTick() {
        long tempTimeTicker = GetCurrentTimeTick();

        while (tempTimeTicker <= _LastTimeTick) {
            tempTimeTicker = GetCurrentTimeTick();
        }

        return tempTimeTicker;
    }

    @Override
    public long nextId() {
        synchronized (_SyncLock) {
            return _IsOverCost ? NextOverCostId() : NextNormalId();
        }
    }
}

File: Go/source/yitidgen/core/snowWorkerM1.go
package core

import (
	"../contract"
	"math"
	"sync"
	"time"
)

type SnowWorkerM1 struct {
	BaseTime                int64  //基础时间
	WorkerId                uint16 //机器码
	WorkerIdBitLength       byte   //机器码位长
	SeqBitLength            byte   //自增序列数位长
	MaxSeqNumber            uint32 //最大序列数（含）
	MinSeqNumber            uint32 //最小序列数（含）
	TopOverCostCount        uint   //最大漂移次数
	_TimestampShift         byte
	_CurrentSeqNumber       uint32
	_LastTimeTick           int64
	_TurnBackTimeTick       int64
	_TurnBackIndex          byte
	_IsOverCost             bool
	_OverCostCountInOneTerm uint
	_GenCountInOneTerm      uint
	_TermIndex              uint
	sync.Mutex
}

func NewSnowWorkerM1(options *contract.IdGeneratorOptions) contract.ISnowWorker {
	var WorkerIdBitLength byte
	var SeqBitLength byte
	var MaxSeqNumber uint32

	var WorkerId = options.WorkerId

	if options.WorkerIdBitLength == 0 {
		WorkerIdBitLength = 6
	} else {
		WorkerIdBitLength = options.WorkerIdBitLength
	}
	if options.SeqBitLength == 0 {
		SeqBitLength = 6
	} else {
		SeqBitLength = options.SeqBitLength
	}
	if options.MaxSeqNumber > 0 {
		MaxSeqNumber = options.MaxSeqNumber
	} else {
		MaxSeqNumber = uint32(math.Pow(2, float64(options.SeqBitLength))) - 1
	}
	var MinSeqNumber = options.MinSeqNumber
	var TopOverCostCount = options.TopOverCostCount

	var BaseTime int64
	if options.BaseTime != 0 {
		BaseTime = options.BaseTime
	} else {
		BaseTime = 1582136402000
	}
	timestampShift := (byte)(options.WorkerIdBitLength + options.SeqBitLength)
	currentSeqNumber := options.MinSeqNumber
	return &SnowWorkerM1{
		BaseTime:          BaseTime,
		WorkerId:          WorkerId,
		WorkerIdBitLength: WorkerIdBitLength,
		SeqBitLength:      SeqBitLength,
		MaxSeqNumber:      MaxSeqNumber,
		MinSeqNumber:      MinSeqNumber,
		TopOverCostCount:  TopOverCostCount,
		_TimestampShift:   timestampShift,
		_CurrentSeqNumber: currentSeqNumber}
}

func (m1 *SnowWorkerM1) DoGenIdAction(arg *contract.OverCostActionArg) {

}

func (m1 *SnowWorkerM1) BeginOverCostAction(useTimeTick int64) {

}

func (m1 *SnowWorkerM1) EndOverCostAction(useTimeTick int64) {
	if m1._TermIndex > 10000 {
		m1._TermIndex = 0
	}
}

func (m1 *SnowWorkerM1) BeginTurnBackAction(useTimeTick int64) {

}

func (m1 *SnowWorkerM1) EndTurnBackAction(useTimeTick int64) {

}

func (m1 *SnowWorkerM1) NextOverCostId() uint64 {
	currentTimeTick := m1.GetCurrentTimeTick()
	if currentTimeTick > m1._LastTimeTick {
		m1.EndOverCostAction(currentTimeTick)
		m1._LastTimeTick = currentTimeTick
		m1._CurrentSeqNumber = m1.MinSeqNumber
		m1._IsOverCost = false
		m1._OverCostCountInOneTerm = 0
		m1._GenCountInOneTerm = 0
		return m1.CalcId(m1._LastTimeTick)
	}
	if m1._OverCostCountInOneTerm >= m1.TopOverCostCount {
		m1.EndOverCostAction(currentTimeTick)
		m1._LastTimeTick = m1.GetNextTimeTick()
		m1._CurrentSeqNumber = m1.MinSeqNumber
		m1._IsOverCost = false
		m1._OverCostCountInOneTerm = 0
		m1._GenCountInOneTerm = 0
		return m1.CalcId(m1._LastTimeTick)
	}
	if m1._CurrentSeqNumber > m1.MaxSeqNumber {
		m1._LastTimeTick++
		m1._CurrentSeqNumber = m1.MinSeqNumber
		m1._IsOverCost = true
		m1._OverCostCountInOneTerm++
		m1._GenCountInOneTerm++

		return m1.CalcId(m1._LastTimeTick)
	}
	m1._GenCountInOneTerm++
	return m1.CalcId(m1._LastTimeTick)
}

func (m1 *SnowWorkerM1) NextNormalId() uint64 {
	currentTimeTick := m1.GetCurrentTimeTick()
	if currentTimeTick < m1._LastTimeTick {
		if m1._TurnBackTimeTick < 1 {
			m1._TurnBackTimeTick = m1._LastTimeTick - 1
			m1._TurnBackIndex++
			// 每毫秒序列数的前5位是预留位，0用于手工新值，1-4是时间回拨次序
			// 最多4次回拨（防止回拨重叠）
			if m1._TurnBackIndex > 4 {
				m1._TurnBackIndex = 1
			}
			m1.BeginTurnBackAction(m1._TurnBackTimeTick)
		}
		time.Sleep(time.Duration(10) * time.Millisecond)
		return m1.CalcTurnBackId(m1._TurnBackTimeTick)
	}
	// 时间追平时，_TurnBackTimeTick清零
	if m1._TurnBackTimeTick > 0 {
		m1.EndTurnBackAction(m1._TurnBackTimeTick)
		m1._TurnBackTimeTick = 0
	}
	if currentTimeTick > m1._LastTimeTick {
		m1._LastTimeTick = currentTimeTick
		m1._CurrentSeqNumber = m1.MinSeqNumber
		return m1.CalcId(m1._LastTimeTick)
	}
	if m1._CurrentSeqNumber > m1.MaxSeqNumber {
		m1.BeginOverCostAction(currentTimeTick)
		m1._TermIndex++
		m1._LastTimeTick++
		m1._CurrentSeqNumber = m1.MinSeqNumber
		m1._IsOverCost = true
		m1._OverCostCountInOneTerm = 1
		m1._GenCountInOneTerm = 1

		return m1.CalcId(m1._LastTimeTick)
	}
	return m1.CalcId(m1._LastTimeTick)
}

func (m1 *SnowWorkerM1) CalcId(useTimeTick int64) uint64 {
	result := uint64(useTimeTick<<m1._TimestampShift) + uint64(m1.WorkerId<<m1.SeqBitLength) + uint64(m1._CurrentSeqNumber)
	m1._CurrentSeqNumber++
	return result
}

func (m1 *SnowWorkerM1) CalcTurnBackId(useTimeTick int64) uint64 {
	result := uint64(useTimeTick<<m1._TimestampShift) + uint64(m1.WorkerId<<m1.SeqBitLength) + uint64(m1._TurnBackIndex)
	m1._TurnBackTimeTick--
	return result
}

func (m1 *SnowWorkerM1) GetCurrentTimeTick() int64 {
	var millis = time.Now().UnixNano() / 1e6
	return millis - m1.BaseTime
}

func (m1 *SnowWorkerM1) GetNextTimeTick() int64 {
	tempTimeTicker := m1.GetCurrentTimeTick()
	for tempTimeTicker <= m1._LastTimeTick {
		tempTimeTicker = m1.GetCurrentTimeTick()
	}
	return tempTimeTicker
}

func (m1 *SnowWorkerM1) NextId() uint64 {
	m1.Lock()
	defer m1.Unlock()
	if m1._IsOverCost {
		return m1.NextOverCostId()
	} else {
		return m1.NextNormalId()
	}
}
